File: codeforces/round379/E.cc
// Codeforces Round #379 (Div. 2)
// http://codeforces.com/contest/734

// ID: 734E (Anton and Tree)

#include <iostream>
#include <stdio.h>
#include <string.h>
#include <vector>
#include <algorithm>
using namespace std;

const int MAX = 200000 + 1;
int color[MAX], level[MAX];
bool visited[MAX];
vector<int> G[MAX], H[MAX], rootGroup;

void dfs(int u, int rootV) {
	visited[u] = true;
	for(int i=0; i<G[u].size(); i++){
		int v = G[u][i];
		if(!visited[v]){
			if(color[v] != color[u]){
				rootGroup.push_back(v);
				H[rootV].push_back(v);
				H[v].push_back(rootV);
				dfs(v, v);
			}else{
				dfs(v, rootV);
			}
		}
	}
}

void bfs(int u, int lvl) {
	visited[u] = true;
	level[u] = lvl;
	for(int i=0; i<H[u].size(); i++){
		int v = H[u][i];
		if(!visited[v]){
			bfs(v, lvl + 1);
		}
	}
}

void solve() {
	int n;
	cin >> n;
	for(int i=1; i<=n; i++) scanf("%d", color + i);
	for(int i=0; i<n-1; i++){
		int u, v; scanf("%d %d", &u, &v);
		G[u].push_back(v);
		G[v].push_back(u);
	}

	memset(visited, 0, sizeof(visited));
	rootGroup.push_back(1);
	dfs(1, 1);

	memset(visited, 0, sizeof(visited));
	bfs(1, 0);
	int end1 = 1;
	for(int i=0; i<rootGroup.size(); i++){
		if(level[rootGroup[i]] > level[end1]){
			end1 = rootGroup[i];
		}
	}
	memset(visited, 0, sizeof(visited));
	bfs(end1, 0);
	int end2 = 1;
	for(int i=0; i<rootGroup.size(); i++){
		if(level[rootGroup[i]] > level[end2]){
			end2 = rootGroup[i];
		}
	}

	int step = (level[end2] + 1) / 2;
	cout << step << endl;
}

int main() {
	solve();
	return 0;
}

File: codeforces/round379/E.hs
-- Codeforces Round #379 (Div. 2)
-- http://codeforces.com/contest/734

-- ID: 734E (Anton and Tree)

import Control.Exception
import Control.DeepSeq
import Control.Monad
import qualified Data.Sequence as Seq
import Data.Array
import Data.List
import Data.Maybe
import Data.Tuple (swap)
import Data.Foldable (toList)
import qualified Data.IntSet as Set
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BS8

getInt = read `fmap` getLine :: IO Int
getInts = (map parseInt . BS8.words) `fmap` BS.getLine :: IO [Int]
parseInt = fst . fromJust . BS8.readInt
type Graph = Array Int [Int]

main = do
	n <- getInt
	colors <- listArray (1,n) `fmap` getInts
	edges <- replicateM (n-1) $ do
		[u,v] <- getInts
		return (u,v)
	let edges' = edges ++ (map swap edges)
	let g = accumArray (flip (:)) [] (1,n) edges'
	--print $ (diameter h n + 1) `div` 2
	print $ (diameter g colors n + 1) `div` 2

-- this is the problem!!! very slow for large N
preorder :: Graph -> Int -> [(Int,Int)]
preorder g start = toList $ bfs Seq.empty [(start,start)] where
	bfs nodes [] = nodes
	bfs nodes queue = bfs (nodes Seq.>< Seq.fromList queue) (concatMap nexts queue)
	nexts (v,parent) = map (\u -> (u,v)) $ filter (/= parent) (g ! v)

--preorder g start = dfs start start where
--	dfs parent v = (v,parent) : concatMap (dfs v) (filter (/= parent) (g ! v))

levels :: Graph -> Array Int Int -> Int -> Int -> Array Int Int
levels g colors start n = a where
	a = array (1, n) [findLevel v | v <- order]
	findLevel (v, parent)
		| v == parent = (v, 0)
		| colors ! v == colors ! parent = (v, a ! parent)
		| otherwise = (v, (a ! parent) + 1)
	order = preorder g start

maxLevel g colors start n
	= maximumBy (\a b -> compare (snd a) (snd b)) $ assocs $ levels g colors start n

diameter :: Graph -> Array Int Int -> Int -> Int
diameter g colors n = snd $ maxLevel g colors (fst $ maxLevel g colors 1 n) n

{-
preorder :: Graph -> Int -> [(Int,Int)]
preorder g start = dfs Set.empty start start where
	dfs visited parent v = (v,parent) : concatMap (dfs visited' v) nexts where
		visited' = Set.insert v visited
		nexts = filter (`Set.notMember` visited) (g ! v)
-}

{-
compress :: Array Int Int -> Graph -> Int -> Int -> Graph
compress colors g start n = accumArray (flip (:)) [] (1,n) edges' where
	edges = map findRoot $ filter diffColor $ order
	edges' = edges ++ (map swap edges)
	findRoot (v,u) = (rootLink ! v, rootLink ! u)
	rootLink = array (1, n) $ map getRoot $ order
	getRoot (v, parent)
		| v == parent || diffColor (v, parent) = (v, v)
		| otherwise = (v, rootLink ! parent)
	diffColor (v,u) = colors ! v /= colors ! u
	order = preorder g start
levels :: Graph -> Int -> Int -> Array Int Int
levels g start n = a where
	a = array (1,n) [findLevel v | v <- preorder g start]
	--a = accumArray (flip const) (-1) (1,n) [findLevel v | v <- preorder g start]
	findLevel (v,parent)
		| v == parent = (v,0)
		| otherwise = (v, (a ! parent) + 1)
--maxLevel g start n = maximumBy (\a b -> compare (snd a) (snd b)) $ assocs $ levels g start n
-}

